src/Microsoft.ApplicationInsights/Managed/Shared/Extensibility/Implementation/Tracing/DiagnosticsTelemetryModule.cs

```diff
@@ -161,6 +274,8 @@ private void Dispose(bool managed)
                     disposableSender.Dispose();
                 }
 
+                this.HeartbeatProvider = null;
```

Comment:
System.Threading,Timer is disposable, do we need to dispose it as part of HeartbeatProvider?

Reply:
Yes, missed this, thank you.

Reply:
Interesting and much deeper issue than I had anticipated. Since the HeartbeatProvider class contains all the data it needs and doesn't require state sharing, I think it is ok to have it naturally dispose on its own with regular GC. However, within the HeartbeatProvider itself we still need to ensure that the timer is disposed in a friendly way that won't affect consumers of our SDK negatively.

Definitely take a look at how I implemented this and give me any feedback you can, I'm interested in getting this right for sure!

Reply:
In addition to wiring it up within HeartBeat provider I would consider making HeartBeat disposable to propagate dispose directly as soon as it's called on `DiagnosticsTelemetryModule` (it's already implementing `IDisposable` anyway to dispose event listener and senders - it won't hurt, as I hope, to dispose HeartBeat unless you envision usage scenarios of this provider that won't dispose it or won't need to dispose it).